torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -53,17 +53,66 @@ Message deserialize(std::istream& is) {
   return Message(std::move(payload), std::move(tensors), type, id);
 }
 
+std::unordered_map<std::string, int> collectNames(
+    const std::string& workerName, std::shared_ptr<c10d::ProcessGroup>& pg) {
+
+  const auto worldSize = pg->getSize();
```

Comment:
I think we can simplify this function if we enforce a limit on the max length of workerName (ex: 128/256 chars). With this, we can just initialize tensors with this max length and just do one all gather to collect all the worker names. All the worker names would be null terminated and hence we don't even need to collect their lengths first. std::string() would do the right thing and just read up to the first null terminated character when we just pass in char * with no length.

Reply:
Sure, but all the optimization we do here will probably gone after we address #23232, as we will call into that instead. Do you still prefer adding the temporary optimization here? LMK, I will make the change.

Reply:
I think let's just add the temporary optimization for now since its not clear when we might address https://github.com/pytorch/pytorch/issues/23232. Also, apart from an optimization, it also just makes the code a lot simpler and cleaner.